### ISSUE TITLE
Support for SMT/ATS co-programming :)

### ISSUE THREAD

**Issue by steinwaywhw** [wdblair/ATS-Postiats-contrib#2]
Hi @wdblair, 

It would be nice if we have a function taking a string and directly printing it at the beginning of generated SMT code. The reason is that when there are things not currently supported by the ATS Z3 api, we can at least manually code it. 

For example, in order to support multiset/bag, I modified the original `set` definition in Z3 to the followings, 

``` smt

;; An encoding of various set-theoretic operators 
;; using the CAL combinators in Z3.

(define-sort Elt () Int)
(define-sort Set () (Array Elt Int))
(declare-fun min (Int Int) Int)
(declare-fun max (Int Int) Int)
(assert (forall ((x Int) (y Int)) (= (min x y) (ite (< x y) x y))))
(assert ())
(define-fun smt_set_emp () Set ((as const Set) 0))
(define-fun smt_set_mem ((x Elt) (s Set)) Bool (> (select s x) 0))
(define-fun smt_set_add ((s Set) (x Elt)) Set  (store s x (+ (select s x) 1)))
(define-fun smt_set_del ((s Set) (x Elt)) Set  (store s x (ite (> (select s x) 0) (- (select s x) 1) 0)))
(define-fun smt_set_cap ((s1 Set) (s2 Set)) Set ((_ map min) s1 s2))
(define-fun smt_set_cup ((s1 Set) (s2 Set)) Set ((_ map (+ (Int Int) Int)) s1 s2))
;;(define-fun smt_set_com ((s Set)) Set ((_ map not) s))
;;(define-fun smt_set_dif ((s1 Set) (s2 Set)) Set (smt_set_cap s1 (smt_set_com s2)))
;;(define-fun smt_set_sub ((s1 Set) (s2 Set)) Bool (= smt_set_emp (smt_set_dif s1 s2)))

(define-const e Set smt_set_emp)
(define-const a Set (smt_set_add (smt_set_add e 1) 1))
(define-const b Set (smt_set_add e 1))
(push)
(assert (not (= (smt_set_cap b a) b)))
(check-sat)

(pop)
```

But encoding this in ATS is not as easy as just printing them as string from ATS to SMT. What do you think?

`Z3_parse_smtlib2_string` may help

**Comment by steinwaywhw** [comment]
Please see 13ebd53a9334ef7070096634f4f4d0bb7d812a35. But not sure where to put a call to these functions. Namely, where is the main entry for generating constraints?

**Comment by steinwaywhw** [comment]
Closed as in 7b69aff908aa55ee9c0b35f47dbeb0221561969c
